### PR TITLE
refactor: 13274 Migrated `MerkleHederaState` to `swirlds-platform-core`.

### LINKED ISSUE #13274: Refactor `HederaState` to `MerkleStateRoot`
Previously, `HederaState` interface resided in `hedera-node` and it represented Hedera app specifically. Now, it moved to `platform-sdk`, which may be a platform for other applications. So, the interface should be renamed to more generic `MerkleState`. 

Also, `MerkleHederaState` is still a part of `hedera-app` and it shouldn't be. It should migrate to `swirlds-platform-core`

### PR DESCRIPTION
**Pull request by imalygin**
**Description**:

PR highlights:

- renamed  `MerkleHederaState` to `MerkleStateRoot`
- migrated `com.hedera.node.app.state.merkle.MerkleTestBase` to `com.swirlds.platform.test.fixtures.state.MerkleTestBase`
- updated dependencies
- changed visibility modifier from package-private to public for `MerkleStateRoot.putServiceStateIfAbsent` and  `MerkleStateRoot.removeServiceState`

**Related issue(s)**:

Fixes #13274 

**Notes for reviewer**:

**Checklist**

- [ ] Documented (Code comments, README, etc.)
- [ ] Tested (unit, integration, etc.)

### CODE REVIEW

File: platform-sdk/swirlds-platform-core/src/main/java/com/swirlds/platform/state/MerkleStateRoot.java [hunk 1]
```diff
@@ -84,7 +81,7 @@
  * SwirldState}. The Hedera application, after startup, only needs the ability to get {@link
```
Comment (lpetrovic05) [inline platform-sdk/swirlds-platform-core/src/main/java/com/swirlds/platform/state/MerkleStateRoot.java:81]: seems this javadoc needs an update as well

File: platform-sdk/swirlds-platform-core/src/main/java/com/swirlds/platform/state/HederaLifecycles.java [hunk 2]
```diff
@@ -14,10 +14,9 @@
  * limitations under the License.
  */
 
-package com.hedera.node.app.state.merkle;
+package com.swirlds.platform.state;
```
Comment (lpetrovic05) [inline platform-sdk/swirlds-platform-core/src/main/java/com/swirlds/platform/state/HederaLifecycles.java:17]: seems that we should not have a class called Hedera in the platform

**Comment by lpetrovic05** [comment]
Also I noticed that `MerkleHederaState` has some workaround for the constructable registry. I think it would be sufficient to add the `@ConstructableIgnored` annotation and get rid of this workaround.

File: hedera-node/hedera-app/src/main/java/com/hedera/node/app/state/merkle/MerkleSchemaRegistry.java [hunk 3]
```diff
@@ -198,8 +200,8 @@ public void migrate(
         if (isSoOrdered(currentVersion, previousVersion)) {
             throw new IllegalArgumentException("The currentVersion must be at least the previousVersion");
         }
-        if (!(hederaState instanceof MerkleHederaState state)) {
-            throw new IllegalArgumentException("The state must be an instance of MerkleHederaState");
+        if (!(hederaState instanceof MerkleStateRoot state)) {
+            throw new IllegalArgumentException("The state must be an instance of MerkleStateRoot");
```
Comment (hendrikebbers) [inline hedera-node/hedera-app/src/main/java/com/hedera/node/app/state/merkle/MerkleSchemaRegistry.java:204]:
```suggestion
            throw new IllegalArgumentException("The state must be an instance of " + MerkleStateRoot.class.getName());
```

File: hedera-node/hedera-app/src/main/java/com/hedera/node/app/statedumpers/StateDumper.java [hunk 4]
```diff
@@ -123,8 +123,8 @@ public static void dumpModChildrenFrom(
             @NonNull final HederaState hederaState,
             @NonNull final DumpCheckpoint checkpoint,
             @NonNull final Set<MerkleStateChild> childrenToDump) {
-        if (!(hederaState instanceof MerkleHederaState state)) {
-            throw new IllegalArgumentException("Expected a MerkleHederaState");
+        if (!(hederaState instanceof MerkleStateRoot state)) {
+            throw new IllegalArgumentException("Expected a MerkleStateRoot");
```
Comment (hendrikebbers) [inline hedera-node/hedera-app/src/main/java/com/hedera/node/app/statedumpers/StateDumper.java:127]:
```suggestion
            throw new IllegalArgumentException("Expected a " + MerkleStateRoot.class.getSimpleName());
```

File: platform-sdk/swirlds-platform-core/src/main/java/com/swirlds/platform/state/HederaLifecycles.java (hunk 2 above)
Comment (hendrikebbers) [reply platform-sdk/swirlds-platform-core/src/main/java/com/swirlds/platform/state/HederaLifecycles.java:17]: Agree. It should be renamed

File: platform-sdk/swirlds-platform-core/src/main/java/com/swirlds/platform/state/MerkleStateRoot.java (hunk 1 above)
Comment (imalygin) [reply platform-sdk/swirlds-platform-core/src/main/java/com/swirlds/platform/state/MerkleStateRoot.java:81]: Fixed

File: platform-sdk/swirlds-platform-core/src/test/java/com/swirlds/platform/state/MerkleStateRootTest.java [hunk 5]
```diff
@@ -59,9 +60,9 @@
 import org.mockito.junit.jupiter.MockitoExtension;
 
 @ExtendWith(MockitoExtension.class)
-class MerkleHederaStateTest extends MerkleTestBase {
+class MerkleStateRootTest extends MerkleTestBase {
     /** The merkle tree we will test with */
-    private MerkleHederaState hederaMerkle;
+    private MerkleStateRoot hederaMerkle;
```
Comment (litt3) [inline platform-sdk/swirlds-platform-core/src/test/java/com/swirlds/platform/state/MerkleStateRootTest.java:65]: `hederaMerkle` could be renamed

**Review by david-bakin-sl — DISMISSED** [review]
LGTM (not sure why smart-contracts was needed to approve, but here it is)

File: hedera-node/hedera-app/src/main/java/com/hedera/node/app/state/WorkingStateAccessor.java [hunk 6]
```diff
@@ -18,38 +18,38 @@
...
     /**
-     * Returns the working {@link HederaState}.
-     * @return the working {@link HederaState}.
+     * Returns the working {@link MerkleState}.
+     * @return the working {@link MerkleState}.
      */
     @Nullable
-    public HederaState getHederaState() {
-        return hederaState;
+    public MerkleState getHederaState() {
```
Comment (litt3) [inline hedera-node/hedera-app/src/main/java/com/hedera/node/app/state/WorkingStateAccessor.java:43]: getter name should be updated

File: hedera-node/hedera-app/src/main/java/com/hedera/node/app/state/WorkingStateAccessor.java [hunk 7]
```diff
@@ -18,38 +18,38 @@
...
     /**
-     * Sets the working {@link HederaState}.
-     * @param hederaState the working {@link HederaState}.
+     * Sets the working {@link MerkleState}.
+     * @param merkleState the working {@link MerkleState}.
      */
-    public void setHederaState(HederaState hederaState) {
-        requireNonNull(hederaState);
-        this.hederaState = hederaState;
+    public void setHederaState(MerkleState merkleState) {
```
Comment (litt3) [inline hedera-node/hedera-app/src/main/java/com/hedera/node/app/state/WorkingStateAccessor.java:51]: setter name should be updated

File: hedera-node/hedera-app/src/main/java/com/hedera/node/app/state/merkle/MerkleSchemaRegistry.java [hunk 8]
```diff
@@ -262,7 +264,7 @@ private RedefinedWritableStates applyStateDefinitions(
             @NonNull final Schema schema,
             @NonNull final Configuration configuration,
             @NonNull final Metrics metrics,
-            @NonNull final MerkleHederaState hederaState) {
+            @NonNull final MerkleStateRoot hederaState) {
```
Comment (litt3) [inline hedera-node/hedera-app/src/main/java/com/hedera/node/app/state/merkle/MerkleSchemaRegistry.java:267]: arg name should be updated

File: hedera-node/hedera-app/src/main/java/com/hedera/node/app/state/recordcache/RecordCacheImpl.java [hunk 9]
```diff
@@ -353,7 +353,7 @@ public List<TransactionRecord> getRecords(@NonNull final AccountID accountID) {
     private WritableStates getWritableState() {
         final var hederaState = workingStateAccessor.getHederaState();
```
Comment (litt3) [inline hedera-node/hedera-app/src/main/java/com/hedera/node/app/state/recordcache/RecordCacheImpl.java:354]: var name can be updated

File: platform-sdk/swirlds-platform-core/src/main/java/com/swirlds/platform/state/MerkleStateLifecycles.java [hunk 10]
```diff
@@ -52,7 +51,7 @@ public interface HederaLifecycles {
      * @param platformState the working state of the platform
      * @param state the working state of the network
      */
```
Comment (litt3) [inline platform-sdk/swirlds-platform-core/src/main/java/com/swirlds/platform/state/MerkleStateLifecycles.java:53]: this javadoc still mentions "Hedera"

File: platform-sdk/swirlds-platform-core/src/main/java/com/swirlds/platform/state/MerkleStateLifecycles.java [hunk 11]
```diff
@@ -52,7 +51,7 @@ public interface HederaLifecycles {
...
-    void onHandleConsensusRound(@NonNull Round round, @NonNull PlatformState platformState, @NonNull HederaState state);
+    void onHandleConsensusRound(@NonNull Round round, @NonNull PlatformState platformState, @NonNull MerkleState state);
 
     /**
      * Called when the platform is initializing the Hedera network state.
```
Comment (litt3) [inline platform-sdk/swirlds-platform-core/src/main/java/com/swirlds/platform/state/MerkleStateLifecycles.java:57]: javadoc still mentions hedera, x2

File: platform-sdk/swirlds-platform-core/src/main/java/com/swirlds/platform/state/MerkleStateLifecycles.java [hunk 12]
```diff
@@ -79,12 +78,12 @@ void onStateInitialized(
      * @param context the current platform context
      */
     void onUpdateWeight(
```
Comment (litt3) [inline platform-sdk/swirlds-platform-core/src/main/java/com/swirlds/platform/state/MerkleStateLifecycles.java:80]: javadoc still mentions Hedera

File: platform-sdk/swirlds-platform-core/src/main/java/com/swirlds/platform/state/MerkleStateRoot.java [hunk 13]
```diff
@@ -135,7 +132,7 @@ public class MerkleHederaState extends PartialNaryMerkleInternal
     /**
      * The callbacks for Hedera lifecycle events.
```
Comment (litt3) [inline platform-sdk/swirlds-platform-core/src/main/java/com/swirlds/platform/state/MerkleStateRoot.java:133]: javadoc needs to be updated

Comment (imalygin) [reply platform-sdk/swirlds-platform-core/src/main/java/com/swirlds/platform/state/MerkleStateRoot.java:121]: Done

File: platform-sdk/swirlds-platform-core/src/main/java/com/swirlds/platform/state/MerkleStateLifecycles.java (hunk 10 above)
Comment (imalygin) [reply platform-sdk/swirlds-platform-core/src/main/java/com/swirlds/platform/state/MerkleStateLifecycles.java:53]: Removed mentions of Hedera

File: hedera-node/hedera-app/src/main/java/com/hedera/node/app/state/recordcache/RecordCacheImpl.java (hunk 9 above)
Comment (imalygin) [reply hedera-node/hedera-app/src/main/java/com/hedera/node/app/state/recordcache/RecordCacheImpl.java]: Done

File: hedera-node/hedera-app/src/main/java/com/hedera/node/app/state/merkle/MerkleSchemaRegistry.java (hunk 8 above)
Comment (imalygin) [reply hedera-node/hedera-app/src/main/java/com/hedera/node/app/state/merkle/MerkleSchemaRegistry.java]: Done

File: hedera-node/hedera-app/src/main/java/com/hedera/node/app/state/WorkingStateAccessor.java (hunk 7 above)
Comment (imalygin) [reply hedera-node/hedera-app/src/main/java/com/hedera/node/app/state/WorkingStateAccessor.java]: Done

File: hedera-node/hedera-app/src/main/java/com/hedera/node/app/state/WorkingStateAccessor.java (hunk 6 above)
Comment (imalygin) [reply hedera-node/hedera-app/src/main/java/com/hedera/node/app/state/WorkingStateAccessor.java]: Done

File: platform-sdk/swirlds-platform-core/src/test/java/com/swirlds/platform/state/MerkleStateRootTest.java (hunk 5 above)
Comment (imalygin) [reply platform-sdk/swirlds-platform-core/src/test/java/com/swirlds/platform/state/MerkleStateRootTest.java]: Done

File: hedera-node/hedera-app/src/main/java/com/hedera/node/app/records/BlockRecordInjectionModule.java [hunk 14]
```diff
@@ -77,11 +77,11 @@ public static BlockRecordManager provideBlockRecordManager(
             @NonNull final ConfigProvider configProvider,
             @NonNull final WorkingStateAccessor state,
             @NonNull final BlockRecordStreamProducer streamFileProducer) {
-        final var hederaState = state.getHederaState();
-        if (hederaState == null) {
+        final var merkleState = state.getMerkleState();
+        if (merkleState == null) {
             throw new IllegalStateException("Hedera state is null");
```
[the remark this answers is not in the document]
Comment (imalygin) [reply hedera-node/hedera-app/src/main/java/com/hedera/node/app/records/BlockRecordInjectionModule.java:82]: Done

**Review by litt3 — APPROVED** [review]

**Review by lpetrovic05 — APPROVED** [review]

File: hedera-node/hedera-app/src/main/java/com/hedera/node/app/Hedera.java [hunk 15]
```diff
@@ -279,12 +279,12 @@ public Hedera(
...
         try {
-            // And the factory for the MerkleHederaState class id must be our constructor
+            // And the factory for the MerkleStateRoot class id must be our constructor
             // FUTURE WORK: https://github.com/hashgraph/hedera-services/issues/11773
             constructableRegistry.registerConstructable(new ClassConstructorPair(
-                    MerkleHederaState.class, () -> new MerkleHederaState(new HederaLifecyclesImpl(this))));
+                    MerkleStateRoot.class, () -> new MerkleStateRoot(new MerkleStateLifecyclesImpl(this))));
         } catch (final ConstructableRegistryException e) {
-            logger.error("Failed to register MerkleHederaState factory with ConstructableRegistry", e);
+            logger.error("Failed to register MerkleStateRoot factory with ConstructableRegistry", e);
```
Comment (hendrikebbers) [inline hedera-node/hedera-app/src/main/java/com/hedera/node/app/Hedera.java:287]:
```suggestion
            logger.error("Failed to register " + MerkleStateRoot.class + " factory with ConstructableRegistry", e);
```
This should not block the merge. Only fix this if another change is needed

**Review by hendrikebbers — APPROVED** [review]

File: hedera-node/hedera-app/src/main/java/com/hedera/node/app/Hedera.java (hunk 15 above)
Comment (imalygin) [reply hedera-node/hedera-app/src/main/java/com/hedera/node/app/Hedera.java:287]: sounds good, thank you! I'll make this change a part of the next PR that will follow immediately after this one

**Review by artemananiev — APPROVED** [review]

File: hedera-node/hedera-app/src/main/java/com/hedera/node/app/Hedera.java (hunk 15 above)
Comment (imalygin) [reply hedera-node/hedera-app/src/main/java/com/hedera/node/app/Hedera.java:287]: I fixed it here https://github.com/hashgraph/hedera-services/pull/14356/commits/ddf6a2eb7c9733a6db8c7efa97c05d4c5ed2866e

File: platform-sdk/swirlds-platform-core/src/main/java/com/swirlds/platform/state/MerkleStateRoot.java [hunk 16]
```diff
@@ -96,28 +93,16 @@
...
-     * {@link ConstructableRegistry} that maps {@link #CLASS_ID} to {@link MerkleHederaState} using a lambda that will
-     * create the instancing using the non-default constructor. But the {@link ConstructableRegistry} doesn't
-     * actually register the second mapping! So we will trick it. We will return this bogus class ID if the default
-     * constructor is used, or {@link #CLASS_ID} otherwise.
-     */
-    private static final long DO_NOT_USE_IN_REAL_LIFE_CLASS_ID = 0x0000deadbeef0000L;
-
     //    private static final long CLASS_ID = 0x2de3ead3caf06392L;
     // Uncomment the following class ID to run a mono -> modular state migration
     // NOTE: also change class ID of ServicesState
```
Comment (mhess-swl) [inline platform-sdk/swirlds-platform-core/src/main/java/com/swirlds/platform/state/MerkleStateRoot.java:108]: Not a blocker, but these lines can be removed. They're irrelevant after the 0.49 (modularization) release.

**Review by mhess-swl — APPROVED** [review]
LGTM!

File: platform-sdk/swirlds-platform-core/src/main/java/com/swirlds/platform/state/MerkleStateRoot.java (hunk 16 above)
Comment (imalygin) [reply platform-sdk/swirlds-platform-core/src/main/java/com/swirlds/platform/state/MerkleStateRoot.java:108]: will do in the follow-up PR

**Review by david-bakin-sl — APPROVED** [review]
LGTM

**Review by nathanklick — APPROVED** [review]

**Review by rbarkerSL — APPROVED** [review]

**Review by tinker-michaelj — APPROVED** [review]
LGTM, though as discussed out-of-band I think a `StateRoot` interface with a `MerkleStateRoot` implementation is more in the spirit of the States API design.

**Pull request merged**